src/Microsoft.AspNetCore.Mvc.Core/Problem.cs

```diff
@@ -0,0 +1,82 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Mvc
+{
+    /// <summary>
+    /// A machine-readable format for specifying errors in HTTP API responses based on https://tools.ietf.org/html/rfc7807.
+    /// </summary>
+    public class Problem : Dictionary<string, object>
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SerializableError"/> class.
+        /// </summary>
+        public Problem()
+            : base(StringComparer.OrdinalIgnoreCase)
+        {
+        }
+
+        /// <summary>
+        /// A URI reference [RFC3986] that identifies the problem type. This specification encourages that, when
+        /// dereferenced, it provide human-readable documentation for the problem type
+        /// (e.g., using HTML [W3C.REC-html5-20141028]).  When this member is not present, its value is assumed to be
+        /// "about:blank".
+        /// </summary>
+        public string Type
+        {
+            get => GetValue<string>(nameof(Type));
+            set => Add(nameof(Type), value);
+        }
+
+        /// <summary>
+        /// A short, human-readable summary of the problem type.It SHOULD NOT change from occurrence to occurrence
+        /// of the problem, except for purposes of localization(e.g., using proactive content negotiation; 
+        /// see[RFC7231], Section 3.4).
+        /// </summary>
+        public string Title
+        {
+            get => GetValue<string>(nameof(Title));
+            set => Add(nameof(Title), value);
+        }
+
+        /// <summary>
+        /// The HTTP status code([RFC7231], Section 6) generated by the origin server for this occurrence of the problem. 
+        /// </summary>
+        public int? Status
+        {
+            get => GetValue<int?>(nameof(Status));
+            set => Add(nameof(Status), value);
+        }
+
+        /// <summary>
+        /// A human-readable explanation specific to this occurrence of the problem. 
+        /// </summary>
+        public string Detail
+        {
+            get => GetValue<string>(nameof(Detail));
+            set => Add(nameof(Detail), value);
+        }
+
+        /// <summary>
+        /// A URI reference that identifies the specific occurrence of the problem.It may or may not yield further information if dereferenced.
+        /// </summary>
+        public string Instance
+        {
+            get => GetValue<string>(nameof(Instance));
+            set => Add(nameof(Instance), value);
+        }
+
+        private TValue GetValue<TValue>(string key)
+        {
+            if (TryGetValue(key, out var value))
+            {
+                return (TValue)value;
+            }
+
+            return default(TValue);
+        }
```

Comment:
This is kinda a minefield. The 'dictionary' is mutable, so it's just as sensible to add the status code as a string, which would throw if you use the getter.

Should we just write a custom serializer for this?